### DISCUSSION TITLE
Pod Affinity for a multi tenant environment

### DISCUSSION THREAD

**Discussion by llamageddon83** [apache/airflow#29847] [General] [answered]
Hello everyone. 

We are running Airflow on Kubernetes in an enterprise environment with multiple tenants. Each tenant/team has its own sets of DAGs. We are running into a noisy neighbor problem where one misbehaving DAG can clog up the resources. 

Can Airflow via some configuration to execute DAGs on a particular set of pods? For example, DAGs owned by Team A, should always be scheduled on pods labeled team A, DAGs owned by Team B should always run on pods labeled as Team B, and so on.

This would allow us to run a single instance of Airflow while solving the noisy neighbor problem and isolating DAGs based on team ownership. 

Thank you.

**Comment by llamageddon83** [comment]
Looking at configurations available in `KubernetesPodOperator`, that may do the above. If anyone has run into this before, any feedback would be appreciated.

**Reply by potiuk** [reply]
Running fully multi-tenant airflow is not currently possible - we are in the process of multi-tenancy implementation and you can watch the summit talk from last year https://www.youtube.com/watch?v=nBNbVhz7rhE (and we are in-progress of implementing it).

There are no simple solutions for now. You can implement CI and code review rules to enforce some rules on your DAGs. For exmple you can force that all tasks in dags from a given tenant use custom executor_config (see https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/kubernetes.html).

**Reply by llamageddon83** [reply]
ooh nice... will definitely play with it.. 🙇

**Comment by hussein-awala** [comment, marked as answer]
If you are using Kubernetes Executor, you can override the pod template and add the affinity for each task (or for each dag via default args):
```python
PythonOperator(
    task_id="some_task",
    python_callable=some_callable,
    executor_config={
        "pod_override": k8s.V1Pod(
            spec=k8s.V1PodSpec(
                affinity=k8s.V1Affinity(...),
            )
        ),
    },
)
```
But if you are using Celery Executor, this is not possible, and using `KubernetesPodOperator` cannot solve the problem beacuse you need to dockerize all the tasks, and you will not be able to use the providers operators.

**Reply by llamageddon83** [reply]
Oh wow. This should do. Gonna play with it. Thank you 🙇

**Reply by potiuk** [reply]
Nice one!
